### PR TITLE
Amannocci faster json parsing

### PR DESCRIPTION
**Pull request by pmlopes**
Implement a `toBuffer` to all json objects.

This PR extends #1976 and fixes the ip-check that was resolved using the github web interface which does not sign off commits.

### CODE REVIEW

**Comment by amannocci** [comment]
Maybe we can directly write into a buffer ?
```java
  public static Buffer encodeToBuffer(Object obj) throws EncodeException {
    try {
      Buffer buf = Buffer.buffer();
      mapper.writeValue(new ByteBufOutputStream(buf.getByteBuf()), obj);
      return buf;
    } catch (Exception e) {
      throw new EncodeException("Failed to encode as JSON: " + e.getMessage());
    }
  }
```

We need also to remove the method decodeValue based on InputStream.
```java
public static <T> T decodeValue(InputStream stream, Class<T> clazz) throws DecodeException {
```

I will attempt to do a little jmh benchmark to confirm performance improvement before merge.
Refer to #1975

**Comment by amannocci** [comment]
We can add a method with TypeReference too ?
```java
public static <T> T decodeValue(Buffer buf, TypeReference<T> type) throws DecodeException
```

**Comment by pmlopes** [comment]
@amannocci your comments have been added to the PR

**Comment by amannocci** [comment]
@pmlopes @vietj Sounds good for me.
Benchs are available on #1975

File: src/main/java/io/vertx/core/json/Json.java [hunk 1]
```diff
@@ -69,6 +72,16 @@ public static String encode(Object obj) throws EncodeException {
     }
   }
 
+  public static Buffer encodeToBuffer(Object obj) throws EncodeException {
```
Comment (vietj) [inline src/main/java/io/vertx/core/json/Json.java]: missing doc

File: src/main/java/io/vertx/core/json/Json.java [hunk 2]
```diff
@@ -93,6 +106,22 @@ public static String encodePrettily(Object obj) throws EncodeException {
     }
   }
 
+  public static <T> T decodeValue(Buffer buf, TypeReference<T> type) throws DecodeException {
```
Comment (vietj) [inline src/main/java/io/vertx/core/json/Json.java]: missing doc

File: src/main/java/io/vertx/core/json/Json.java (hunk 1 above)
Comment (vietj) [inline src/main/java/io/vertx/core/json/Json.java]: missing doc

File: src/main/java/io/vertx/core/json/Json.java (hunk 2 above)
Comment (vietj) [inline src/main/java/io/vertx/core/json/Json.java]: missing doc

File: src/main/java/io/vertx/core/json/Json.java [hunk 3]
```diff
@@ -93,6 +106,22 @@ public static String encodePrettily(Object obj) throws EncodeException {
...
+    try {
+      return mapper.readValue(new ByteBufInputStream(buf.getByteBuf()), type);
+    } catch (Exception e) {
+      throw new DecodeException("Failed to decode:" + e.getMessage(), e);
+    }
+  }
+
+  public static <T> T decodeValue(Buffer buf, Class<T> clazz) throws DecodeException {
```
Comment (vietj) [inline src/main/java/io/vertx/core/json/Json.java]: missing doc

File: src/main/java/io/vertx/core/json/JsonArray.java [hunk 4]
```diff
@@ -545,6 +554,15 @@ public String encode() {
   }
 
   /**
+   * Encode this JSON object as a string.
+   *
+   * @return the string encoding.
```
Comment (vietj) [inline src/main/java/io/vertx/core/json/JsonArray.java]: the encoded string, string encoding means something else

File: src/main/java/io/vertx/core/json/JsonObject.java [hunk 5]
```diff
@@ -764,6 +773,15 @@ public String encodePrettily() {
   }
 
   /**
+   * Encode this JSON object as a string.
+   *
+   * @return the string encoding.
```
Comment (vietj) [inline src/main/java/io/vertx/core/json/JsonObject.java]: the encoded string

**Comment by jponge** [comment]
To be clear: I do think your changes boost performance, but I'd like a more rigorous evaluation in your benchmarks to know exactly by how much 😉 (and maybe you will just end up with the same figures)

**Comment by amannocci** [comment]
I'm agree with you @jponge. I will try to improve the actual benchmark to be sure.

**Comment by amannocci** [comment]
I have done some modifications on my benchmark as proposed by @jponge .
I get similar results.

```
# Run complete. Total time: 00:27:19

Benchmark                         Mode  Cnt       Score     Error  Units
JsonBenchmark.testNewJsonArray   thrpt  200   52004,058 ± 391,153  ops/s
JsonBenchmark.testNewJsonObject  thrpt  200  103620,960 ± 681,895  ops/s
JsonBenchmark.testOldJsonArray   thrpt  200   40164,809 ± 295,782  ops/s
JsonBenchmark.testOldJsonObject  thrpt  200   78245,268 ± 608,647  ops/s
```

Let me know if I'm wrong.

**Comment by jponge** [comment]
LGTM 👍

**Comment by amannocci** [comment]
LGTM

**Comment by vietj** [comment]
can you rebase the branch so it can be merged ? thanks

**Comment by pmlopes** [comment]
it can be merged, why do we need to rebase?

**Comment by vietj** [comment]
I can  see "This branch cannot be rebased due to conflicts" in github

**Comment by jponge** [comment]
@vietj "Merging can be performed automatically" 😉

**Comment by vietj** [comment]
the UI was not really clear

**Pull request merged**